grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcHealthCheckService.java

```diff
@@ -0,0 +1,286 @@
+                serverHealth.setHealthy(false);
+                watchers.keySet().forEach(StreamObserver::onCompleted);
+                watchers.clear();
+            }
+        });
+    }
+
+    @VisibleForTesting
+    ServingStatus checkServingStatus(@Nullable String serviceName) {
+        if (!isHealthy()) {
+            return ServingStatus.NOT_SERVING;
+        }
+        if (Strings.isNullOrEmpty(serviceName)) {
+            return ServingStatus.SERVING;
+        }
+        final ListenableHealthChecker listenableHealthChecker = grpcServiceHealthCheckers.get(serviceName);
+        if (listenableHealthChecker == null) {
+            return ServingStatus.SERVICE_UNKNOWN;
+        }
+        if (listenableHealthChecker.isHealthy()) {
+            return ServingStatus.SERVING;
+        }
+        return ServingStatus.NOT_SERVING;
+    }
+
+    @VisibleForTesting
+    void changeServerStatus(boolean isHealthy) {
+        serverHealth.setHealthy(isHealthy);
+    }
+
+    private void addServerHealthUpdateListener(List<HealthCheckUpdateListener> updateListeners) {
+        serverHealth.addListener(healthChecker -> {
+            updateListeners.forEach(updateListener -> {
+                try {
+                    updateListener.healthUpdated(healthChecker.isHealthy());
+                } catch (Throwable t) {
+                    logger.warn("Unexpected exception from HealthCheckUpdateListener.healthUpdated():", t);
+                }
+            });
+        });
+    }
+
+    private void setInternalHealthUpdateListener(
+            HealthCheckUpdateListener healthCheckUpdateListener,
+            Collection<ListenableHealthChecker> listenableHealthCheckers) {
+        listenableHealthCheckers.forEach(lhc -> {
+            lhc.addListener(healthChecker -> {
+                try {
+                    healthCheckUpdateListener.healthUpdated(healthChecker.isHealthy());
+                } catch (Throwable t) {
+                    logger.warn("Unexpected exception from HealthCheckUpdateListener.healthUpdated():", t);
+                }
+            });
+        });
+    }
+
+    private HealthCheckUpdateListener provideInternalHealthUpdateListener() {
+        return isHealthy -> {
+            if (updating) {
+                return;
+            }
+            updating = true;
+            try {
+                for (Entry<StreamObserver<HealthCheckResponse>, Entry<String, ServingStatus>> entry
+                        : watchers.entrySet()) {
+                    final ServingStatus previousStatus = entry.getValue().getValue();
+                    final String serviceName = entry.getValue().getKey();
+                    final ServingStatus currentStatus = checkServingStatus(serviceName);
+                    if (currentStatus != previousStatus) {
+                        final StreamObserver<HealthCheckResponse> responseObserver = entry.getKey();
+                        responseObserver.onNext(getHealthCheckResponse(currentStatus));
```

Comment:
I think this line may throw an exception.

1) From a different thread, the request is cancelled and the `responseObserver` is removed from`watchers`
2) Right afterwards, we try to notify the removed `responseObserver` with `onNext` and an exception is thrown
3) The `for` loop is terminated without notifying all watchers

I think some concurrency control method may be necessary (or we could `try..catch` if we're lazy, but I think this might cause unexpected problems later on)

Reply:
No need to worry about this comment if we follow the `synchronized` approach 😅